compiler/rustc_hir_typeck/src/lib.rs

```diff
@@ -259,6 +259,21 @@ fn typeck_with_inspect<'tcx>(
 
     let typeck_results = fcx.resolve_type_vars_in_body(body);
 
+    // Handle potentially region dependent goals, see `InferCtxt::in_hir_typeck`.
+    if let None = fcx.infcx.tainted_by_errors() {
+        for obligation in fcx.take_hir_typeck_potentially_region_dependent_goals() {
+            let obligation = fcx.resolve_vars_if_possible(obligation);
+            if obligation.has_non_region_infer() {
+                bug!("unexpected inference variable after writeback: {obligation:?}");
+            }
+            fcx.register_predicate(obligation);
+        }
+        fcx.select_obligations_where_possible(|_| {});
```

Comment:
```suggestion
        // We must only evaluate the registered goals after they have *all* been added
        // so that the previous assertions are not order dependent. Not that it would be
        // easy to evaluate the goals one by one anyway.
        fcx.select_obligations_where_possible(|_| {});
```

Reply:
do we? :thinking: this does not feel like a meaningful invariant to me